File: sem4/MPP/proiect/appClientServer/CompetitionPersistence/src/main/java/org/example/persistence/repository/ParticipationDBRepository.java
package org.example.persistence.repository;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.example.model.Child;
import org.example.model.Participation;
import org.example.model.Task;
import org.example.persistence.IChildRepository;
import org.example.persistence.IParticipationRepository;
import org.example.persistence.ITaskRepository;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;


public class ParticipationDBRepository implements IParticipationRepository {
    private JdbcUtils dbUtils;
    private IChildRepository ChildRepo;
    private ITaskRepository TaskRepo;

    private static final Logger logger= LogManager.getLogger();

    public ParticipationDBRepository(Properties props, IChildRepository child, ITaskRepository task) {
        logger.info("Initializing ParticipationDBRepository with properties: {} ",props);
        dbUtils=new JdbcUtils(props);
        ChildRepo = child;
        TaskRepo = task;
    }
    @Override
    public Iterable<Participation> findByTask(Task task) {
        logger.traceEntry();
        Connection con= dbUtils.getConnection();
        List<Participation> participations = new ArrayList<>();
        try (PreparedStatement preStmt=con.prepareStatement("select * from participation where id_task = ?")){
            preStmt.setInt(1, task.getId());
            try (ResultSet result= preStmt.executeQuery()){
                while (result.next()){
                    int idP = result.getInt("id_participation");
                    int idC = result.getInt("id_child");
                    int idT = result.getInt("id_task");
                    Child child = ChildRepo.findById(idC);
                    Task task1 = TaskRepo.findById(idT);
                    Participation participation = new Participation(child,task);
                    participation.setId(idP);
                    participations.add(participation);
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        logger.traceExit(participations);
        return participations;
    }

    @Override
    public void add(Participation elem) {
        logger.traceEntry("saving task {} ", elem);
        Connection con= dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into participation (id_child, id_task) values (?, ?)")){
            preStmt.setInt(1, elem.getChild().getId());
            preStmt.setInt(2, elem.getTask().getId());
            int result= preStmt.executeUpdate();
            logger.trace("Saved {} instances", elem);
        }catch (SQLException ex){
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Participation elem) {

    }

    @Override
    public void update(Participation elem, Integer id) {

    }

    @Override
    public Participation findById(Integer id) {
        logger.traceEntry();
        Connection con= dbUtils.getConnection();
        try (PreparedStatement preStmt=con.prepareStatement("select * from participation where id_participation = ?")){
            preStmt.setInt(1, id);
            try (ResultSet result= preStmt.executeQuery()){
                while (result.next()){
                    int idP = result.getInt("id_participation");
                    int idC = result.getInt("id_child");
                    int idT = result.getInt("id_task");
                    Child child = ChildRepo.findById(idC);
                    Task task = TaskRepo.findById(idT);
                    Participation participation = new Participation(child,task);
                    participation.setId(idP);
                    return participation;
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        logger.traceExit();
        return null;
    }

    @Override
    public Iterable<Participation> findAll() {
        logger.traceEntry();
        Connection con= dbUtils.getConnection();
        List<Participation> participations = new ArrayList<>();
        try (PreparedStatement preStmt=con.prepareStatement("select * from participation")){
            try (ResultSet result= preStmt.executeQuery()){
                while (result.next()){
                    int idP = result.getInt("id_participation");
                    int idC = result.getInt("id_child");
                    int idT = result.getInt("id_task");
                    Child child = ChildRepo.findById(idC);
                    Task task = TaskRepo.findById(idT);
                    Participation participation = new Participation(child,task);
                    participation.setId(idP);
                    participations.add(participation);
                }
            }
        }catch (SQLException ex){
            logger.error(ex);
            System.err.println("Error DB " + ex);
        }
        logger.traceExit(participations);
        return participations;
    }
}

File: sem4/MPP/proiectC#/ClientServerApp/persistence/ParticipationDbRepository.cs
using System;
using System.Collections.Generic;
using System.Data;
using log4net;
using model;
namespace repository;

public class ParticipationDbRepository : IParticipationRepository
{
	private static readonly ILog log = LogManager.GetLogger("SortingTaskDbRepository");
	private ChildDbRepository repoC;
	private TaskDbRepository repoT;

	IDictionary<String, string> props;

	public ParticipationDbRepository(IDictionary<String, string> props)
	{
		log.Info("Creating SortingTaskDbRepository ");
		this.props = props;
		repoC = new ChildDbRepository(props);
		repoT = new TaskDbRepository(props);
	}

	public IEnumerable<Participation> findByTask(Task task1)
	{
		log.InfoFormat("Entering findOne with value {0}", task1);
		IDbConnection con = DBUtils.getConnection(props);

		using (var comm = con.CreateCommand())
		{
			comm.CommandText = "select * from participation where id_task=@id";
			IDbDataParameter paramId = comm.CreateParameter();
			paramId.ParameterName = "@id";
			paramId.Value = task1.ID;
			comm.Parameters.Add(paramId);

			IList<Participation> participations = new List<Participation>();
			using (var dataR = comm.ExecuteReader())
			{
				while (dataR.Read())
				{
					int idP = dataR.GetInt32(0);
					int idC = dataR.GetInt32(1);
					int idT = dataR.GetInt32(2);
					Child child = repoC.FindOne(idC);
					Task task = repoT.FindOne(idT);
					Participation participation = new Participation(child, task);
					participation.ID = idP;
					participations.Add(participation);
				}
			}
			return participations;
		}
	}
	

		public Participation FindOne(int id)
		{
			log.InfoFormat("Entering findOne with value {0}", id);
			IDbConnection con = DBUtils.getConnection(props);

			using (var comm = con.CreateCommand())
			{
				comm.CommandText = "select * from participation where id_participation=@id";
				IDbDataParameter paramId = comm.CreateParameter();
				paramId.ParameterName = "@id";
				paramId.Value = id;
				comm.Parameters.Add(paramId);

				using (var dataR = comm.ExecuteReader())
				{
					if (dataR.Read())
					{
						int idP = dataR.GetInt32(0);
						int idC = dataR.GetInt32(1);
						int idT = dataR.GetInt32(2);
						Child child = repoC.FindOne(idC);
						Task task = repoT.FindOne(idT);
						Participation participation = new Participation(child, task);
						participation.ID = idP;
						log.InfoFormat("Exiting findOne with value {0}", participation);
						return participation;
					}
				}
			}
			log.InfoFormat("Exiting findOne with value {0}", null);
			return null;
		}

		public IEnumerable<Participation> FindAll()
		{
			IDbConnection con = DBUtils.getConnection(props);
			IList<Participation> participations = new List<Participation>();
			using (var comm = con.CreateCommand())
			{
				comm.CommandText = "select * from participation";

				using (var dataR = comm.ExecuteReader())
				{
					while (dataR.Read())
					{
						int idP = dataR.GetInt32(0);
						int idC = dataR.GetInt32(1);
						int idT = dataR.GetInt32(2);
						Child child = repoC.FindOne(idC);
						Task task = repoT.FindOne(idT);
						Participation participation = new Participation(child, task);
						participation.ID = idP;
						participations.Add(participation);
					}
				}
			}
			return participations;
		}
		
		public Participation Save(Participation entity)
		{
			var con = DBUtils.getConnection(props);

			using (var comm = con.CreateCommand())
			{
				comm.CommandText =
					"insert into participation(id_child, id_task)  values (@idC, @idT)";

				var paramIdC = comm.CreateParameter();
				paramIdC.ParameterName = "@idC";
				paramIdC.Value = entity.child.ID;
				comm.Parameters.Add(paramIdC);

				var paramIdT = comm.CreateParameter();
				paramIdT.ParameterName = "@idT";
				paramIdT.Value = entity.task.ID;
				comm.Parameters.Add(paramIdT);

				var result = comm.ExecuteNonQuery();
				if (result == 0)
					//throw new RepositoryException("No task added !");
					throw new Exception("No task added !");
				else
				{
					return entity;
				}
			}
		}

		public Participation Delete(int id)
		{
			return null;
		}

		public Participation Update(Participation entity)
		{
			return null;
		}
}